### PR TITLE
model parallel training research

### PR DESCRIPTION
**Pull request by wyli**
### Description
adds a research prototype for model parallel training

### Status
**ready**

### Types of changes

- [x] non Breaking new features
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: research/lamp-automated-model-parallelism/train.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        train_dataset = Dataset(train_images, transform=train_transform)
+        # when bs > 1, the loader assumes that the full image sizes are the same across the dataset
+        train_dataloader = torch.utils.data.DataLoader(train_dataset, num_workers=6, batch_size=bs, shuffle=True)
+    else:
+        train_transform = Compose(
+            [
+                AddChannelDict(keys="image"),
+                CopyItemsd(keys="label", times=1, names="label_fg"),
+                lambda d: d.update({"label_fg": d["label_fg"][1:]}) or d,  # excluding label bg from the sampling mask
+                RandCropByPosNegLabeld(keys=("image", "label"), label_key="label_fg", size=crop_size, num_samples=bs),
```
Comment (wyli) [inline research/lamp-automated-model-parallelism/train.py:91]: looks like `RandCropByPosNegLabeld`'s `label_key` doesn't accept one-hot labels that have both foreground and background channels? here I use a copyitem to implement a workaround, any comments/suggestions @Nic-Ma?

Comment (Nic-Ma) [reply research/lamp-automated-model-parallelism/train.py:91]: Hi @wyli ,

Thanks for our bug report!
I made a quick update: https://github.com/Project-MONAI/MONAI/pull/620, could you please help review it?

Thanks.

Comment (wyli) [reply research/lamp-automated-model-parallelism/train.py]: > Hi @wyli ,
> 
> Thanks for our bug report!
> I made a quick fix: #620, could you please help review it?
> 
> Thanks.

thanks it works

**Comment by wyli** [comment]
> * I update to use deform augmentation for whole image. It works if we use latest Rand3DElasticd.
> * I tested all the jobs and they can run successfully. Thank you!

thanks! I also verified it locally, the commands work properly. 
@Nic-Ma if everything looks ok, please merge this.

File: research/lamp-automated-model-parallelism/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+Wentao Zhu, Can Zhao, Wenqi Li, Holger Roth, Ziyue Xu, and Daguang Xu (2020)
+"LAMP: Large Deep Nets with Automated Model Parallelism for Image Segmentation."
+MICCAI 2020 (Early Accept, paper link: https://arxiv.org/abs/2006.12575)
+
+
+## To run the demo:
+
+### Prerequisites
+- install the latest version of MONAI: `git clone https://github.com/Project-MONAI/MONAI` and `pip install -e .`
```
Comment (Nic-Ma) [inline research/lamp-automated-model-parallelism/README.md:20]: Let's try to update to `pip install monai==0.2.0` later, otherwise if we don't maintain this research example, it may be not compatible with further MONAI versions.

File: research/lamp-automated-model-parallelism/data_utils.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    Args:
+        path: data folder name
+        maskname: target structure names
+    """
+    maskfiles = []
+    for seg in maskname:
+        if os.path.exists(os.path.join(path, "./structures/" + seg + "_crp_v2.npy")):
+            maskfiles.append(os.path.join(path, "./structures/" + seg + "_crp_v2.npy"))
+        else:
+            # print("missing annotation", seg, path.split("/")[-1])
```
Comment (Nic-Ma) [inline research/lamp-automated-model-parallelism/data_utils.py:42]: please remove the commented line.

File: research/lamp-automated-model-parallelism/unet_pipe.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            if isinstance(child, nn.Sequential):
+                for sub_name, sub_child in _flatten(child):
+                    yield f"{name}_{sub_name}", sub_child
+            else:
+                yield name, child
+
+    return nn.Sequential(OrderedDict(_flatten(module)))
+
+
+class DoubleConv(nn.Module):
```
Comment (Nic-Ma) [inline research/lamp-automated-model-parallelism/unet_pipe.py:59]: I feel `DoubleConv` is a common module in many networks, could you please move it to network/blocks?

File: research/lamp-automated-model-parallelism/unet_pipe.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        self.conv = nn.Sequential(
+            Convolution(spatial_dims, in_channels, out_channels, strides=stride, act=act_1, norm=norm_1, bias=False,),
+            Convolution(spatial_dims, out_channels, out_channels, act=act_2, norm=norm_2, conv_only=conv_only),
+        )
+
+    def forward(self, x):
+        return self.conv(x)
+
+
+class UNet(nn.Sequential):
```
Comment (Nic-Ma) [inline research/lamp-automated-model-parallelism/unet_pipe.py:100]: As this UNet implementation is confusing with our internal UNet, could you please help change the class name to some name specific to this research?

File: research/lamp-automated-model-parallelism/README.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+Head and Neck CT dataset
+
+Please download and unzip the images into `./data` folder.
+
+- `HaN.zip`: https://drive.google.com/file/d/1A2zpVlR3CkvtkJPvtAF3-MH0nr1WZ2Mn/view?usp=sharing
+```bash
+unzip HaN.zip;  # unzip
+```
+
+Please find more details of the dataset at https://github.com/wentaozhu/AnatomyNet-for-anatomical-segmentation.git
```
[the remark this answers is not in the document]
Comment (wentaozhu) [reply research/lamp-automated-model-parallelism/README.md:37]: + Actually, there is no need to know the details of the dataset if someone only wants to run the job or use the automated model parallelism.
+ As you know, data is the most important part, especially in medical image analysis. If someone is interested in knowing the details of the dataset, he/she can dig deep into it from the link.
+ The contribution of the dataset is from this work https://github.com/wentaozhu/AnatomyNet-for-anatomical-segmentation.git

File: research/lamp-automated-model-parallelism/unet_pipe.py (hunk 4 above)
Comment (wyli) [reply research/lamp-automated-model-parallelism/unet_pipe.py:59]: yes, it is common, this block is only suitable for the two cases used in this research. we'll need another PR to introduce more args properly and make it generic

File: research/lamp-automated-model-parallelism/README.md (hunk 2 above)
Comment (wyli) [reply research/lamp-automated-model-parallelism/README.md:20]: sounds good

File: research/lamp-automated-model-parallelism/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (wyli) [reply research/lamp-automated-model-parallelism/README.md:37]: I'm not aware of the internal discussion, but if it's necessary we can remove it later.
